kobo/apps/trackers/models.py

```diff
@@ -0,0 +1,64 @@
+from django.contrib.auth.models import User
+from django.db import models
+from django.db.models import F, Q, Value
+from django.db.models.constraints import UniqueConstraint
+from django.db.models.expressions import Func, RawSQL
+from django.db.models.functions import Cast
+from django.db.models.signals import post_delete
+
+
+class MonthlyNLPUsageCounter(models.Model):
+    year = models.IntegerField()
+    month = models.IntegerField()
+    user = models.ForeignKey(User, related_name='users', on_delete=models.DO_NOTHING)
+    asset = models.ForeignKey("kpi.asset", null=True, on_delete=models.CASCADE)
+    counters = models.JSONField(default=dict)
+
+    class Meta:
+        constraints = [
+            UniqueConstraint(fields=['year', 'month', 'user', 'asset'],
+                             name='unique_with_asset'),
+            UniqueConstraint(fields=['year', 'month', 'user'],
+                             condition=Q(asset=None),
+                             name='unique_without_asset')
+        ]
+        indexes = [
+            models.Index(fields=('year', 'month', 'user')),
+        ]
+
+    @classmethod
+    def update_catch_all_counter_on_delete(cls, sender, instance, **kwargs):
+        criteria = dict(
+            year=instance.year,
+            month=instance.month,
+            user=instance.user,
+            asset=None,
+        )
+        # make sure an instance exists with `asset = NULL`
+        cls.objects.get_or_create(**criteria)
+        # add the count for the project being deleted to the null-xform
+        # instance, atomically!
+        counter_keys = instance.counters.keys()
+        for service in counter_keys:
+            # Updating this way because other methods were messy and less reliable
+            # than using the rawSQL method
+            sql = f"""
+                jsonb_set(
+                    counters,
+                    '{{{service}}}',
+                    (COALESCE(counters->>'{service}','0')::int + {instance.counters[service]})::text::jsonb
+                )
+            """
+            MonthlyNLPUsageCounter.objects.filter(**criteria).update(
+                counters=RawSQL(sql, [])
+            )
```

Comment:
What about calling your utility function directly? 

```python 
def update_catch_all_counter_on_delete(cls, sender, instance, **kwargs):
     criteria = dict(
           year=instance.year,
           month=instance.month,
           user_id=instance.user_id,
           asset=None,
     )
     # make sure an instance exists with `asset = NULL`
     counter = cls.objects.get_or_create(**criteria)

    services = instance.counters.keys()
    for service in instances.counters.keys():
        update_nlp_counter(service, amount=instances.counters[service], user_id=instance.user_id, counter_id=counter.pk)
...

def update_nlp_counter(
        service: str,
        amount: int,
        user_id: int,
        asset_id: Optional[int] = None,
        counter_id: Optional[int] = None,
):
    if not counter_id:
        date = datetime.today()
        criteria = dict(
            year=date.year,
            month=date.month,
            user_id=user_id,
            asset_id=asset_id,
       )
      # Ensure the counter for the month exists first
      counter =  MonthlyNLPUsageCounter.objects.get_or_create(**criteria)
      counter_id = counter.pk
      
      MonthlyNLPUsageCounter.objects.filter(pk=counter_id).update(
          counters=IncrementValue('counters', keyname=service, increment=amount)
      )
```

Reply:
I avoided calling the utility function here because it caused a circular import

Reply:
@JacquelineMorrissette, let's talk about this.